source/agora/network/Client.d

```diff
@@ -227,7 +230,20 @@ public class NetworkClient
             break;
 
         case Preimage:
-            this.attemptRequest!(API.postPreimage, Throw.No)(event.preimage);
+            // Clean up the `last_preimages` periodically
+            // Note: It clears `last_preimages` everyday based on the CoinNet.
+            if (event.preimage.height % Height(144) == 0)
+                this.last_preimages.clear();
+
+            // We do not post the preimage if it is already posted in order to
+            // prevent calls of `postPreimage` from flooding
+            if (event.preimage.utxo !in this.last_preimages ||
+                event.preimage.height > this.last_preimages[event.preimage.utxo])
+            {
+                auto height = this.attemptRequest!(API.postPreimage, Throw.No)(event.preimage);
+                if (height >= event.preimage.height)
+                    this.last_preimages[event.preimage.utxo] = event.preimage.height;
+            }
```

Comment:
```suggestion
            if (event.preimage.height > this.last_preimages.get(event.preimage.utxo, Height(0)))
            {
                auto height = this.attemptRequest!(API.postPreimage, Throw.No)(event.preimage);
                if (height >= event.preimage.height)
                    this.last_preimages[event.preimage.utxo] = event.preimage.height;
            }
```

Reply:
The `get` function is not `nothrow`. Isn't it better to leave as it is?